packages/newsroom-signup/src/Newsroom.tsx

```diff
@@ -368,6 +360,37 @@ class NewsroomComponent extends React.Component<NewsroomProps & DispatchProp<any
     );
   }
 
+  private determineInitialStep(savedStep: STEP): STEP {
+    let currentStep = savedStep;
+    if (currentStep === STEP.APPLIED) {
+      // Not a real step, see its description above
+      currentStep--;
+    }
+
+    if (qs.parse(document.location.search.substr(1)).purchased) {
+      // Just been redirected back from token purchase
+      currentStep = STEP.TOKENS;
+    }
+
+    if (this.props.grantRequested && typeof this.props.grantApproved !== "boolean") {
+      // Waiting on grant
+      currentStep = STEP.PROFILE_GRANT;
+    }
+
+    currentStep = this.backtrackSteps(currentStep);
+
+    return currentStep;
+  }
+
+  /** Handle situation where user used nav to jump too far ahead in step process before we put in disable checks - they should be backtracked back to where they need to be. */
+  private backtrackSteps(step: STEP): STEP {
```

Comment:
does this have to be recursive, can't we just say `if there is no address and your current step is past launch contract then go back to the contract launch section` ?

Reply:
That would only handle the case where they clicked too far ahead before making a contract, but people could be anywhere. They could have started on the newsroom profile and immediately clicked on Apply To Registry nav step without having filled in charter, or they could have created contract but tried to skip tutorial, etc. Doing it recursively we get to reuse the logic in `getDisabled` in a granular way.